File: C++/384. Shuffle an Array.cpp
class Solution {
public:
    vector<int>a;
    Solution(vector<int>& nums) {
        a=nums;
    }
    
    vector<int> reset() {
        return a;
    }
    
    vector<int> shuffle() {
        vector<int>x(a.begin(),a.end());
        for(int i=0;i<a.size();i++)
        swap(x[i],x[rand()%a.size()]);
        return x;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * vector<int> param_1 = obj->reset();
 * vector<int> param_2 = obj->shuffle();
 */

File: JAVA/384. Shuffle an Array.java
class Solution {
    private int[] original;
    private Random rand;

    public Solution(int[] nums) {
        original = nums.clone();
        rand = new Random();
    }

    public int[] reset() {
        return original.clone();
    }

    public int[] shuffle() {
        int[] shuffled = original.clone();
        for (int i = 0; i < shuffled.length; i++)
            swap(shuffled, i, rand.nextInt(shuffled.length));
        return shuffled;
    }

    private void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}
